types/block.go

```diff
@@ -56,6 +56,20 @@ func CalculateCoinbase(height BlockHeight) Currency {
 	return NewCurrency64(base).Mul(SiacoinPrecision)
 }
 
+// CalculateNumSiacoins calculates the number of siacoins in circulation at a
+// given height.
+func CalculateNumSiacoins(height BlockHeight) Currency {
+	deflationBlocks := BlockHeight(InitialCoinbase - MinimumCoinbase)
+	avgDeflationSiacoins := CalculateCoinbase(0).Add(CalculateCoinbase(height)).Div(NewCurrency64(2))
+	if height <= deflationBlocks {
+		deflationSiacoins := avgDeflationSiacoins.Mul(NewCurrency64(uint64(height + 1)))
+		return deflationSiacoins
+	}
+	deflationSiacoins := avgDeflationSiacoins.Mul(NewCurrency64(uint64(deflationBlocks + 1)))
+	trailingSiacoins := NewCurrency64(uint64(height - deflationBlocks)).Mul(CalculateCoinbase(height))
+	return deflationSiacoins.Add(trailingSiacoins)
```

Comment:
The more complicated way is computationally more efficient. I don't want to be doing 500,000 bignum operations every time we are counting coins (once the blockchain gets that large). Even today, 25,000 bignum operations is going to be worth optimizing.

The test for this function verifies the result against the simpler calculation.